pyro/distributions/conjugate.py

```diff
@@ -0,0 +1,64 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+from torch.distributions import constraints
+from torch.distributions.utils import broadcast_all
+
+from pyro.distributions.torch import Beta, Binomial
+from pyro.distributions.torch_distribution import TorchDistribution
+
+
+def _log_beta(x, y):
+    return torch.lgamma(x) + torch.lgamma(y) - torch.lgamma(x + y)
+
+
+class BetaBinomial(TorchDistribution):
+    arg_constraints = {'concentration1': constraints.positive, 'concentration0': constraints.positive,
+                       'total_count': constraints.nonnegative_integer}
+    support = Binomial.support
+
+    def __init__(self, concentration1, concentration0, total_count=1, validate_args=None):
+        concentration1, concentration0, total_count = broadcast_all(
+            concentration1, concentration0, total_count)
+        self._beta = Beta(concentration1, concentration0)
+        self.total_count = total_count
+        super(BetaBinomial, self).__init__(self._beta._batch_shape, validate_args=validate_args)
+
+    @property
+    def concentration1(self):
+        return self._beta.concentration1
+
+    @property
+    def concentration0(self):
+        return self._beta.concentration0
+
+    def expand(self, batch_shape, _instance=None):
+        new = self._get_checked_instance(BetaBinomial, _instance)
+        batch_shape = torch.Size(batch_shape)
+        new._beta = self._beta.expand(batch_shape)
+        new.total_count = self.total_count.expand_as(new._beta.concentration0)
+        super(BetaBinomial, new).__init__(batch_shape, validate_args=False)
+        new._validate_args = self._validate_args
+        return new
+
+    def sample(self, sample_shape=()):
+        probs = self.Beta.sample(sample_shape)
+        return Binomial(probs).sample()
```

Comment:
Though I don't know if using Conjugate distributions will be helpful at all in HMC context, I would like to suggest a way to query collapsed variables: we return here a sample with a specific attribute `_collapsed_sample` which stores the sample from Beta distribution (which is `probs`). A specific method `poutine.get_uncollapsed_trace` will be used instead of `poutine.get_trace` when users want to get the samples from Beta distribution. In that method, we add an additional node for this collapsed site. WDYT?